reflect: add `Value.Clear`; support `anytype->interface{}`, `Slice->(*)Array` in `Value.Convert`, fix `Copy`

ben-krieger opened the pull request:
Fixes #4554

This PR adds a few things to the reflect package.

1. Add `Value.Clear`
2. Convert anything to `interface{}`, i.e. `reflect.ValueOf(42).Convert(reflect.TypeFor[interface{}]())`
3. Convert slices to arrays, i.e. `reflect.ValueOf([]int{1, 2, 3}).Convert(reflect.TypeFor[[3]int]())`
4. Convert slices to array pointers... because after doing 3, I felt like I had to
5. Fixes `reflect.Copy` when the source is an array

The motivation was to allow compiling the CBOR library at https://github.com/fido-device-onboard/go-fdo/tree/tinygo-server-support/cbor. However, this PR can be split into smaller ones if desired.

Commit:
reflect: add Value.Clear; support anytype->interface{}, Slice->(*)Array in Value.Convert

dgryski commented on src/reflect/value.go:
```diff
@@ -689,6 +689,27 @@ func (v Value) Cap() int {
+	case Slice:
+		hdr := (*sliceHeader)(v.value)
+		elemSize := v.typecode.underlying().elem().Size()
+		for i := uintptr(0); i < hdr.len; i++ {
```
Shouldn't this just be a single `memzero` call instead of one call for each element in the slice?

ben-krieger replied:
Updated.

dgryski reviewed (commented):
Just noticed an issue, but I'll do a more complete review Thursday.

dgryski commented on src/reflect/type.go:
```diff
@@ -970,10 +976,18 @@ func (t *rawType) FieldAlign() int {
 // of type u.
 func (t *rawType) AssignableTo(u Type) bool {
-	if t == u.(*rawType) {
+	if u := u.(*rawType); t == u || t.underlying() == u || t == u.underlying() {
```
`underlying()` returns the type itself if it's not a named type; so you should just be able to use `t.underlying() == u.underlying()` (or use `t == u` first if you want to potentially avoid the function calls).

ben-krieger replied:
That's actually subtly different and I'm not sure what the correct behavior is.

Currently, I'd say that these two types below are not assignable to one another, though either is assignable to `[]byte` or vice versa.

```go
type Named1 []byte
type Named2 []byte
```

I'm going to look at what's correct in the language spec.

ben-krieger replied:
Yes, your suggestion is the correct behavior. Updated.

dgryski commented on src/reflect/type.go:
```diff
@@ -970,10 +976,18 @@ func (t *rawType) FieldAlign() int {
 		return true
 	}
 
+	if u.Kind() == Array && t.Kind() == Slice {
```
These additions make our `reflect` package differ from upstream:
```
~/go/src/github.com/dgryski/refcvt $ tinygo run main.go
true
true
~/go/src/github.com/dgryski/refcvt $ go run main.go
false
false
~/go/src/github.com/dgryski/refcvt $ cat main.go
package main

import "reflect"

func main() {

	s := make([]byte, 10)
	var hw [5]byte

	st := reflect.TypeOf(s)
	hwt := reflect.TypeOf(hw)
	hwtp := reflect.TypeOf(&hw)

	println(st.AssignableTo(hwt))
	println(st.AssignableTo(hwtp))
}
```

ben-krieger replied:
Hmmm... So the correct change would have been to update `Value.Set` to check `AssignableTo || CanConvert`?

ben-krieger replied:
Updated and now your example returns false false for both.

dgryski commented on src/reflect/value.go:
```diff
@@ -1262,6 +1286,15 @@ func convertOp(src Value, typ Type) (Value, bool) {
 		}, true
 	}
 
+	if rtype := typ.(*rawType); rtype.Kind() == Interface && rtype.NumMethod() == 0 {
```
Do you have a test that triggers this?

ben-krieger replied:
`TestConvertToEmptyInterface`

dgryski commented on src/reflect/value.go:
```diff
@@ -1302,25 +1335,47 @@ func convertOp(src Value, typ Type) (Value, bool) {
+			if src.typecode.elem() == rtype.elem() && rtype.Len() <= src.Len() {
+				return Value{
+					typecode: rtype,
+					value:    (*sliceHeader)(src.value).data,
```
Pretty sure you need to add the `valueFlagIndirect` here to the set of flags, otherwise we'll miss the extra pointer load when we need to get the value.

```
~/go/src/github.com/dgryski/refcvt $ go run main.go
706050403020100 c000102020
~/go/src/github.com/dgryski/refcvt $ tinygo run main.go
10b2514b0 10b2514b0
~/go/src/github.com/dgryski/refcvt $ cat main.go
package main

import (
	"encoding/binary"
	"fmt"
	"reflect"
	"unsafe"
)

func main() {

	s := make([]byte, 8)
	var arr [8]byte

	for i := range s {
		s[i] = byte(i)
	}

	arrT := reflect.TypeOf(arr)
	sV := reflect.ValueOf(s)

	a := sV.Convert(arrT).Interface().([8]byte)

	ptr := binary.LittleEndian.Uint64(a[:])
	fmt.Printf("%x %x\n", ptr, unsafe.SliceData(s))
}
```

ben-krieger replied:
Thank you. I'm fumbling in the dark with the flags, because it's hard to come up with tests to see if they're set correctly. The behavior of when a value is addressable, etc is quite confusing.

ben-krieger replied:
Updated. Here's my test program:

```go
//go:build ignore

package main

import (
	"fmt"
	"reflect"
)

func main() {
	s := make([]byte, 8)
	for i := range s {
		s[i] = byte(i)
	}

	a := reflect.ValueOf(s).Convert(reflect.TypeFor[[8]byte]()).Interface().([8]byte)
	b := reflect.ValueOf(s).Convert(reflect.TypeFor[*[8]byte]()).Interface().(*[8]byte)

	fmt.Printf("&a: %p\n&b: %p\n&s: %p\n", a[:], b[:], s)
	fmt.Printf("a: %v\nb: %v\ns: %v\n", a[:], (*b)[:], s)
}
```

Results:

```console
$ go run ./test.go
&a: 0xc00009a048
&b: 0xc00009a040
&s: 0xc00009a040
a: [0 1 2 3 4 5 6 7]
b: [0 1 2 3 4 5 6 7]
s: [0 1 2 3 4 5 6 7]

$ tinygo run ./test.go
&a: 0x7f07aca106d0
&b: 0x7f07aca106b0
&s: 0x7f07aca106b0
a: [0 1 2 3 4 5 6 7]
b: [0 1 2 3 4 5 6 7]
s: [0 1 2 3 4 5 6 7]
```

dgryski commented on src/reflect/value_test.go:
```diff
@@ -624,6 +648,123 @@ func TestConvert(t *testing.T) {
+	_ = v.Interface().(interface{}).(map[string]string)
+}
+
+func TestClearSlice(t *testing.T) {
```
Also need `TestClearMap`

dgryski commented on src/reflect/value_test.go:
```diff
@@ -1,6 +1,7 @@
 package reflect_test
```
Are there any tests in `all_test.go` you can uncomment?

ben-krieger replied:
I did not know to look there. I'll check!

ben-krieger replied:
I don't think any can be uncommented without significant additions. The most obvious candidate is the conversion tests, but this would require implementing `ConvertibleTo`.

ben-krieger commented:
Thanks for the thorough review @dgryski! I'll spend some time this weekend fixing it up and adding more tests.

Commit:
reflect: fix Copy of non-pointer array with size > 64bits

dgryski commented:
`testdata/reflect.go` is still failing.  That's Bad because the desired output is generated by upstream Go, so that indicates incompatibilities.

ben-krieger commented:
> `testdata/reflect.go` is still failing. That's Bad because the desired output is generated by upstream Go, so that indicates incompatibilities.

Got it. So it seems that arrays in some cases have become addrable and settable, when they shouldn't. Let me see if I can figure out why...

dgryski commented:
It's almost certainly the flag changes for `indirect` support.  You probably need some `RO` bits too.

ben-krieger commented:
> It's almost certainly the flag changes for `indirect` support. You probably need some `RO` bits too.

Tried RO, but couldn't get it working, so I reverted the change that broke it. (I was also uncomfortable with that code, because I didn't understand the underlying mechanism that distinguished between arrays > 64 bits in size.) However, the test I added for `reflect.Copy` is still failing.

Any thoughts on the proper way to fix #4554?

ben-krieger commented:
> It's almost certainly the flag changes for `indirect` support. You probably need some `RO` bits too.

@dgryski I think this is ready for (hopefully the last) review. All tests are passing now and the fix for #4554 matches a lot of other code within the reflect package that check either indirect OR size > uintptr.

I don't see any more low-hanging fruit for test coverage, but let me know what you think. Thanks!

dgryski commented on src/reflect/value.go:
```diff
@@ -1061,8 +1083,8 @@ func (it *MapIter) Next() bool {
 	v.checkRO()
-	if !x.typecode.AssignableTo(v.typecode) {
-		panic("reflect: cannot set")
+	if !x.typecode.AssignableTo(v.typecode) && !x.typecode.ConvertibleTo(v.typecode) {
```
Should this be `||` instead of `&&` ?

ben-krieger replied:
According to the go docs:

> As in Go, x's value must be assignable to v's type and must not be derived from an unexported field.

So this change shouldn't be needed... However, I was not getting correct behavior matching Go without it. 

I'll see if I can add some tests and update AssignableTo with whatever implicit conversions are allowed, instead. 

But yes, the && was correct because it was not A and not B, meaning not (A or B), i.e fail if not assignable or convertible.

dgryski replied:
Thanks.  I think I must've missed one of the `!`'s when I skimmed this.

dgryski commented:
One quick comment.  I'll do a full review tomorrow (Monday) and then hopefully we can get this merged.  Thanks for all your work on this!

ben-krieger commented on src/reflect/type.go:
```diff
@@ -995,6 +1001,10 @@ func (t *rawType) AssignableTo(u Type) bool {
 		return true
 	}
 
+	if t.underlying() == u.(*rawType).underlying() && (!t.isNamed() || !u.(*rawType).isNamed()) {
```
See https://go.dev/ref/spec#Assignability

If the underlying types are the same, only one can be named.

dgryski reviewed (approved):
LGTM

dgryski commented:
@aykevl Any comments?

aykevl commented:
Took a look, looks good to me.
